src/llmq/quorums.cpp

```diff
@@ -381,12 +452,198 @@ CQuorumCPtr CQuorumManager::GetQuorum(Consensus::LLMQType llmqType, const CBlock
+            request.SetError(nError);
+            CDataStream ssResponse(SER_NETWORK, pFrom->GetSendVersion(), request, body);
+            g_connman->PushMessage(pFrom, CNetMsgMaker(pFrom->GetSendVersion()).Make(NetMsgType::QDATA, ssResponse));
+        };
+
+        {
+            LOCK2(cs_main, cs_data_requests);
+            auto key = std::make_pair(pFrom->verifiedProRegTxHash, false);
+            auto it = mapQuorumDataRequests.find(key);
+            if (it == mapQuorumDataRequests.end()) {
+                it = mapQuorumDataRequests.emplace(key, request).first;
+            } else if(it->second.IsExpired()) {
+                it->second = request;
+            } else {
+                error("Request limit exceeded", true, 25);
+            }
+        }
+
+        if (Params().GetConsensus().llmqs.count(request.GetLLMQType()) == 0) {
+            sendQDATA(CQuorumDataRequest::Errors::QUORUM_TYPE_INVALID);
+            return;
+        }
+
+        const CBlockIndex* pQuorumIndex{nullptr};
+        {
+            LOCK(cs_main);
+            pQuorumIndex = LookupBlockIndex(request.GetQuorumHash());
+        }
+        if (pQuorumIndex == nullptr) {
+            sendQDATA(CQuorumDataRequest::Errors::QUORUM_BLOCK_NOT_FOUND);
+            return;
+        }
+
+        const CQuorumCPtr pQuorum = GetQuorum(request.GetLLMQType(), pQuorumIndex);
+        if (pQuorum == nullptr) {
+            sendQDATA(CQuorumDataRequest::Errors::QUORUM_NOT_FOUND);
+            return;
+        }
+
+        CDataStream ssResponseData(SER_NETWORK, pFrom->GetSendVersion());
+
+        if (request.GetDataMask() & CQuorumDataRequest::QUORUM_VERIFICATION_VECTOR) {
+
+            if (!pQuorum->quorumVvec) {
+                sendQDATA(CQuorumDataRequest::Errors::QUORUM_VERIFICATION_VECTOR_MISSING);
+                return;
+            }
+
+            ssResponseData << *pQuorum->quorumVvec;
+        }
+
+        if (request.GetDataMask() & CQuorumDataRequest::ENCRYPTED_CONTRIBUTIONS) {
+
+            int memberIdx = pQuorum->GetMemberIndex(request.GetProTxHash());
+            if (memberIdx == -1) {
+                sendQDATA(CQuorumDataRequest::Errors::MASTERNODE_IS_NO_MEMBER);
+                return;
+            }
+
+            std::vector<CBLSIESEncryptedObject<CBLSSecretKey>> vecEncrypted;
+            if (!quorumDKGSessionManager->GetEncryptedContributions(request.GetLLMQType(), pQuorumIndex, pQuorum->qc.validMembers, request.GetProTxHash(), vecEncrypted)) {
+                sendQDATA(CQuorumDataRequest::Errors::ENCRYPTED_CONTRIBUTIONS_MISSING);
+                return;
+            }
+
+            ssResponseData << vecEncrypted;
+        }
+
+        sendQDATA(CQuorumDataRequest::Errors::NONE, ssResponseData);
+        return;
+    }
+
+    if (strCommand == NetMsgType::QDATA) {
+
+        if (!fMasternodeMode || pFrom == nullptr || (pFrom->verifiedProRegTxHash.IsNull() && !pFrom->qwatch)) {
```

Comment:
A node running with `-watchquorums` should process `QDATA` too
```suggestion
        bool isWatching = gArgs.GetBoolArg("-watchquorums", llmq::DEFAULT_WATCH_QUORUMS);
        if ((!fMasternodeMode && !isWatching) || pFrom == nullptr || (pFrom->verifiedProRegTxHash.IsNull() && !pFrom->qwatch)) {
```

Reply:
See 5c6b2e1, dropped the `llmq` namespace prefix and prefixed the bool (just because i like booleans a bit more with the `f` :D)

Reply:
hmm... you "fixed" the wrong branch with 5c6b2e1, no? it should be `QDATA`, not `QGETDATA`

Reply:
We probably need tests for this :D

Reply:
ops 🙈

Reply:
Okaaay, fixed it in 7fe5f31 and added tests for it in e49a9a9 + see few other commits i pushed